src/gui/GuiRunner.cc

```diff
@@ -115,11 +120,48 @@ GuiRunner::GuiRunner(const std::string &_worldName)
   QPointer<QTimer> timer = new QTimer(this);
   connect(timer, &QTimer::timeout, this, &GuiRunner::UpdatePlugins);
   timer->start(33);
+
+  this->dataPtr->controlService = "/world/" + _worldName + "/control";
+
+  ignition::gui::App()->findChild<
+      ignition::gui::MainWindow *>()->installEventFilter(this);
 }
 
 /////////////////////////////////////////////////
 GuiRunner::~GuiRunner() = default;
 
+/////////////////////////////////////////////////
+bool GuiRunner::eventFilter(QObject *_obj, QEvent *_event)
+{
+  if (_event->type() == ignition::gui::events::WorldControl::kType)
+  {
+    auto worldControlEvent =
+      reinterpret_cast<gui::events::WorldControl *>(_event);
+    if (worldControlEvent)
+    {
+      msgs::WorldControlState req;
+      req.mutable_world_control()->CopyFrom(
+          worldControlEvent->WorldControlInfo());
+
+      // If a user presses play, send the GUI's ECM state to the server
+      // in case the GUI's ECM was modified while paused
+      if (worldControlEvent->Play())
```

Comment:
Can you check that the GUI `step` button also triggers an ECM update? I think that is missing.

Reply:
Yeah, I missed that case - thanks for pointing it out. I think the easiest way to check for play/step is to update the GUI event in https://github.com/ignitionrobotics/ign-gui/pull/306 to have a `Pause` method instead of `Play`. Then, if calling `Pause` returns false, we know that the ECM needs to be updated. Giving the new `WorldControl` GUI event a `Pause` method instead of `Play` also makes it align better with the `WorldControl` message type (`msgs::WorldControl` has a `paused` field). I'll go ahead and make this update, unless you're opposed to this change in any way.

Reply:
The `step` case has been handled in https://github.com/ignitionrobotics/ign-gazebo/pull/1109/commits/3ca5d9dd72626b72209c422464cfb65e92d1eafc. It looks like there's some "interesting" behavior when pressing step now. Whenever I press step while paused, the GUI usually shows that it's playing for a moment, and then goes back to paused (see below). I believe this is due to the extra time it takes for the event approach to be processed, compared to the original service approach (I don't see this GUI behavior when using the service). I'm not sure if this is necessarily a "problem", but I wanted to point it out.

![step](https://user-images.githubusercontent.com/42042756/139778552-e3ce72b5-83e0-4b73-b35e-7a2da46af76c.gif)

Reply:
Yes, I'm seeing this as well. I was able to get the play button into an odd state, where simulation was running but the play button showed the play symbol (>).

Reply:
Clicking the step button rapidly can also cause simulation to start running.

Reply:
Okay, so after debugging, I found why the behaviors above were occurring. Here's the program flow when you press step while paused:
1. When step is pressed while simulation is paused, the server has to set its paused state to `false` so that the step can be processed: https://github.com/ignitionrobotics/ign-gazebo/blob/35a5772ece24cf394cf4d03d95bf0502af31f90e/src/SimulationRunner.cc#L1218
2. Once the paused state is set to false, the step occurs, and world stats are published. So, in the world stats message that's published, paused is false: https://github.com/ignitionrobotics/ign-gazebo/blob/35a5772ece24cf394cf4d03d95bf0502af31f90e/src/SimulationRunner.cc#L415-L418
3. On the `ign-gui` side, the `WorldControl` plugin receives this message and sees that `paused` is false. But, the `WorldControl` plugin is currently paused since step occurred while paused, so the plugin thinks that it needs to be updated to start playing: https://github.com/ignitionrobotics/ign-gui/blob/054d02c87b72db9a80998899d45b372a0cfdac9c/src/plugins/world_control/WorldControl.cc#L245-L264
4. The server goes through its next update: this time, [it resets its paused state to `true`](https://github.com/ignitionrobotics/ign-gazebo/blob/35a5772ece24cf394cf4d03d95bf0502af31f90e/src/SimulationRunner.cc#L844), and has no `WorldControl` msgs to process. So, when the server publishes its next stats message, `paused` is now true.
5. `ign-gui`'s `WorldControl` (which is now in a playing state because of the previous stats message) processes this message, and sees that paused is true. So, now the opposite of step 3 happens, and the `WorldControl` GUI plugin goes back to a paused state.

To work around the issue outlined above, I have added a `step` key to the stats message whenever a message is published that corresponds to a step (see https://github.com/ignitionrobotics/ign-gazebo/pull/1109/commits/cabdc966701dc9fbf72444115d118f52c0d8f6e2). The `WorldControl` GUI plugin has also been updated to ignore any incoming stats messages that have this header/key combination (see https://github.com/ignitionrobotics/ign-gui/pull/306/commits/d7e2c4f31360608f14fe1f1b40fdcec16f7f90fd). This now resolves the issues related to pressing step while paused.

Once strange thing that I noticed is that when using the service approach instead of the new event approach (so, using `<use_event>false</use_event` in the gui config), if step is pressed while paused, the server will still publish a stats message with pause set to false, but when `ign-gui` processes this message, it seems to think that pause in the message is true. When using the event approach, `ign-gui` thinks pause is false when processing the message. So, I am not sure why there's a difference between the two approaches, but the fix I mentioned above ensures proper behavior for both the event and service approach.